extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -238,34 +253,158 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
     return getAmpAdRenderOutsideViewport(this.element) || 3;
   }
 
-  /** @override */
+  /** @override
+      @return {!Promise|undefined}.
+  */
   buildCallback() {
     super.buildCallback();
     this.identityTokenPromise_ = Services.viewerForDoc(this.getAmpDoc())
       .whenFirstVisible()
       .then(() =>
         getIdentityToken(this.win, this.getAmpDoc(), super.getConsentPolicy())
       );
-    this.autoFormat_ = this.element.getAttribute('data-auto-format') || '';
 
-    if (this.isResponsive_()) {
-      // Attempt to resize to the correct height. The width should already be
-      // 100vw, but is fixed here so that future resizes of the viewport don't
-      // affect it.
-      const viewportSize = this.getViewport().getSize();
-      return this.attemptChangeSize(
-        AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
-          this.autoFormat_,
-          viewportSize,
-          this.element
-        ),
-        viewportSize.width
-      ).catch(() => {});
+    return this.maybeUpgradeToResponsive().then(() => {
+      this.autoFormat_ = this.element.getAttribute('data-auto-format') || '';
+
+      if (this.isResponsive_()) {
+        // Attempt to resize to the correct height. The width should already be
+        // 100vw, but is fixed here so that future resizes of the viewport don't
+        // affect it.
+        const viewportSize = this.getViewport().getSize();
+        return this.attemptChangeSize(
+          AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
+            this.autoFormat_,
+            viewportSize,
+            this.element
+          ),
+          viewportSize.width
+        ).catch(() => {});
+      }
+
+      // This should happen last, as some diversion criteria rely on some of the
+      // preceding logic (specifically responsive logic).
+      this.divertExperiments();
+      this.maybeAddSinglePassExperiment();
+    });
+  }
+
+  /**
+   * Upgrades the ad unit to responsive if there is an opt-in setting in localstorage.
+   * See https://github.com/ampproject/amphtml/issues/23568 for design.
+   * @return {!Promise} a promise that resolves when any upgrade is complete.
+   */
+  maybeUpgradeToResponsive() {
```

Comment:
actually for code health can you put all new logic in a separate file. this file is ~700 lines of code already.

Reply:
I'm not really sure it'd get easier to follow if we just move some methods to a new file, I think it'd be preferable to have some sort of logical organization.

If the aim is to reduce the file size, I think I'd suggest moving out static methods to a helper class, in particular since those also make up for a lot of the test cases. WDYT? I'm happy to send a PR.

Reply:
I was thinking of something like this:

```
export class ResponsiveAd {
  maybeUpgradeToResponsive() {};
  ...
} 
```